extensions/AWS/S3/S3Storage.cs

```diff
@@ -0,0 +1,346 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Microsoft.Extensions.Logging;
+using Microsoft.KernelMemory.Diagnostics;
+
+namespace Microsoft.KernelMemory.DocumentStorage.S3;
+
+public class S3Storage : IDocumentStorage, IDisposable
+{
+    private readonly AmazonS3Client _client;
+    private readonly string _bucketName;
+    private readonly ILogger<S3Storage> _log;
+
+    public S3Storage(
+        S3Config config,
+        ILogger<S3Storage>? log = null)
+    {
+        this._log = log ?? DefaultLogger<S3Storage>.Instance;
+
+        switch (config.Auth)
+        {
+            case S3Config.AuthTypes.AccessKey:
+                {
+                    this.ValidateAccessKey(config.AccessKey);
+                    this.ValidateSecretKey(config.SecretKey);
+
+                    this._client = new AmazonS3Client(
+                        awsAccessKeyId: config.AccessKey,
+                        awsSecretAccessKey: config.SecretKey,
+
+                        clientConfig: new AmazonS3Config
+                        {
+                            ServiceURL = config.CustomHost + "/" + config.BucketName,
+                            LogResponse = true
+                        }
+                    );
+                    break;
+                }
+
+            default:
+                this._log.LogCritical("Authentication type '{0}' undefined or not supported", config.Auth);
+                throw new DocumentStorageException($"Authentication type '{config.Auth}' undefined or not supported");
+        }
+
+        if (string.IsNullOrEmpty(config.BucketName))
+        {
+            var msg = $"Bucket name '{config.BucketName}' undefined or not supported";
+            this._log.LogCritical(msg);
+            throw new DocumentStorageException(msg);
+        }
+
+        this._bucketName = config.BucketName;
+    }
+
+    public void Dispose()
+    {
+        this._client?.Dispose();
```

Comment:
Not sure if it makes sense making it disposable. I also added IDisposable to the class:
``` 
public class S3Storage : IDocumentStorage, IDisposable
```

Reply:
you'll need it only if you have some internal disposable references. If `_client` is a disposable object, that looks good